### ISSUE TITLE
GraphQLError request failed with status code 401

### ISSUE THREAD

**Issue by cc-bojan** [aws-amplify/amplify-cli#3485]
**Which Category is your question related to?**
Auth / GraphQL

**What AWS Services are you utilizing?**
AWS AppSync

**Provide additional details e.g. code snippets**

In our ```schema.graphql```, we added ```@function``` directive to the field inside the ```Query``` type, in order to invoke our ```recoverUsername``` lambda function:

``` graphql
type Query {
  recoverUsername(email: String): String
  @function(name: "recoverUsername-${env}")
  @aws_iam
}
```
It works when I execute this query in the AWS AppSync console by selecting IAM auth provider, but when I try to call it from the app with: 
``` js
API.graphql({
  query: recoverUsername,
  variables: {
    email,
  },
  authMode: 'AWS_IAM',
});
```
``` js
aws-exports.js

const awsmobile = {
  aws_appsync_authenticationType: 'AWS_IAM',
  ...
}
```
I get `GraphQLError request failed with status code 401`

We are using older version of `aws-amplify` (1.2.4), but I also tried with latest version, same error.

**Comment by kaustavghosh06** [comment]
@cc-bojan What does your `amplify/backend/backend-config.json` look like? Have you set IAM as an auth type?

**Comment by cc-bojan** [comment]
> @cc-bojan What does your `amplify/backend/backend-config.json` look like? Have you set IAM as an auth type?

```
    "api": {
        "fooApi": {
            "service": "AppSync",
            "providerPlugin": "awscloudformation",
            "output": {
                "authConfig": {
                    "additionalAuthenticationProviders": [
                        {
                            "authenticationType": "AWS_IAM"
                        }
                    ],
                    "defaultAuthentication": {
                        "authenticationType": "AMAZON_COGNITO_USER_POOLS",
                        "userPoolConfig": {
                            "userPoolId": "authFoo"
                        }
                    }
                }
            }
        },
        ...
```

**Comment by attilah** [comment]
@cc-bojan when you're logged into the AWS Console you're probably using an administrator role credentials that does not need to have a specific policy to invoke AppSync mutations, that's the probable reason why it works.

If you use the AppSync service supported native directives like `@aws_iam` the Amplify CLI will not generate any policies for you for the Auth and Unauth roles, so when from the client application you call the Amplify service and specifying IAM as authentication mode, it will use the AWS credentials of the currently logged in user, and that user's role (Auth or Unauth) has no attached policies defined that would grant access for AppSync service to invoke that Lambda function.

Could you please confirm that the above is correct? And if it is correct create a policy and verify that it is working?

**Comment by cc-bojan** [comment]
@attilah You were right. I attached: `arn:aws:appsync:eu-central-1:<account-id>:apis/<api-id>/types/Query/fields/recoverUsername` - Resource to the existing `UnauthRolePolicy`, and now it's working. Thanks.

**Comment by ohsik** [comment]
@cc-bojan Can you please tell me how you attached the policy?

I tried adding below to `CustomResources.json` but still getting 401 error.
```
"Parameters": {
		"authRoleName": {
			"Type": "String"
		},
		"unauthRoleName": {
			"Type": "String"
		}
......
}
```
and

```
			"AuthRolePolicy": {
			"Type": "AWS::IAM::Policy",
			"Properties": {
				"PolicyName": "appsync-authrole-policy-custom",
				"Roles": [
						{
							"Ref": "authRoleName"
						}
					],
				"PolicyDocument": {
					"Version": "2012-10-17",
					"Statement": [
							{
							"Effect": "Allow",
							"Action": ["appsync:GraphQL"],
							"Resource": [
									{
										"Fn::Sub": [
											"arn:aws:appsync:${AWS::Region}:${AWS::AccountId}:apis/${apiId}/types/${typeName}/fields/${fieldName}",
											{
											"apiId": {
												"Ref": "AppSyncApiId"
											},
												"typeName": "Mutation",
												"fieldName": "customMutation"
											}
										]
									}
								]
							}
						]
					}
				}
			},

			"UnauthRolePolicy": {
				"Type": "AWS::IAM::Policy",
				"Properties": {
					"PolicyName": "appsync-unauthrole-policy-custom",
					"Roles": [
							{
								"Ref": "unauthRoleName"
							}
						],
					"PolicyDocument": {
						"Version": "2012-10-17",
						"Statement": [
								{
								"Effect": "Allow",
								"Action": ["appsync:GraphQL"],
								"Resource": [
										{
											"Fn::Sub": [
												"arn:aws:appsync:${AWS::Region}:${AWS::AccountId}:apis/${apiId}/types/${typeName}/fields/${fieldName}",
												{
												"apiId": {
													"Ref": "AppSyncApiId"
												},
													"typeName": "Mutation",
													"fieldName": "customMutation"
												}
											]
										}
									]
								}
							]
						}
					}
				},
``

and my schema.graphql
```
type Mutation {
     customMutation(username: String!, expoNotificationToken: String!): AWSJSON @function(name: "customMutation-${env}") @aws_iam @aws_cognito_user_pools
     customMutation2(username: String!, expoNotificationToken: String!): AWSJSON @function(name: "customMutation2-${env}")
}
```

**Comment by marcusharjani** [comment]
Newb but hopefully this is helpful:

I was getting a 401 error with GraphQL using AWS Amplify with a GraphQL API.

With my authentication method `./src/aws-exports.js` uses `API_KEY` as `aws_appsync_authenticationType`.

After some digging, I realized in `./amplify/backend/backend-config.json` the API key by default expires in seven days (something I thoughtlessly clicked through when spinning up the API with `amplify add api`). 

So I went about updating this. I'm sure there are command line solutions to updating the API key but I went into my AppySync console > API > name of my API > Settings and added a new API key. 

I then updated `./amplify/backend/backend-config.json` with the API description e.g., `            "apiKeyConfig": {
              "description": "yourNewApiKey"
            }`
and updated `./aws-exports.js` with the new API key `    "aws_appsync_apiKey": "newkeyinfo"
`.

**Comment by marcuscodesok** [comment]
Newb but hopefully this is helpful:

I was getting a 401 error with GraphQL using AWS Amplify with a GraphQL API.

With my authentication method `./src/aws-exports.js` uses `API_KEY` as `aws_appsync_authenticationType`.

After some digging, I realized in `./amplify/backend/backend-config.json` the API key by default expires in seven days (something I thoughtlessly clicked through when spinning up the API with `amplify add api`). 

So I went about updating this. I'm sure there are command line solutions to updating the API key but I went into my AppySync console > API > name of my API > Settings and added a new API key. 

I then updated `./amplify/backend/backend-config.json` with the API description e.g., `            "apiKeyConfig": {
              "description": "yourNewApiKey"
            }`
and updated `./aws-exports.js` with the new API key `    "aws_appsync_apiKey": "newkeyinfo"
`.
